plugins/timertop/timermodel.cpp

```diff
@@ -317,37 +540,175 @@ QMap<int, QVariant> TimerModel::itemData(const QModelIndex &index) const
+        m_freeTimersInfo.clear();
+        endRemoveRows();
+    }
+
+    triggerPushChanges();
+}
+
+void TimerModel::applyChanges(const GammaRay::TimerIdInfoHash &changes)
+{
+    QSet<TimerId> handledIds;
+    QVector<QPair<int, int>> dataChangedRanges; // pair of first/last
+    QSet<int> qtimerIds;
+
+    // QQmlTimer / QTimer
+    for (int i = 0; i < m_sourceModel->rowCount(); ++i) {
+        const QModelIndex idx = index(i, 0);
+        const TimerId id = TimerId(static_cast<QObject *>(idx.internalPointer()));
+        const auto cit = changes.constFind(id);
+        auto it = m_timersInfo.find(id);
+
+        if (it == m_timersInfo.end()) {
+            it = m_timersInfo.insert(id, TimerIdInfo());
+        }
+
+        if (cit != changes.constEnd()) {
+            handledIds << id;
+            it.value() = cit.value();
+
+            if (dataChangedRanges.isEmpty() || dataChangedRanges.last().second != i - 1) {
+                dataChangedRanges << qMakePair(i, i);
+            } else {
+                dataChangedRanges.last().second = i;
+            }
+        }
+
+        // Remember QTimer's id for later
+        if (id.type() == TimerId::QTimerType && it.value().timerId != -1) {
+            qtimerIds << it.value().timerId;
+        }
+    }
+
+    // Existing free timers
+    for (auto it = m_freeTimersInfo.begin(), end = m_freeTimersInfo.end(); it != end; ++it) {
+        const TimerId id((*it).timerId);
+        const auto cit = changes.constFind(id);
+
+        if (cit != changes.constEnd()) {
+            const int i = m_sourceModel->rowCount() + std::distance(m_freeTimersInfo.begin(), it);
+            handledIds << id;
+            (*it) = cit.value();
+
+            if (dataChangedRanges.isEmpty() || dataChangedRanges.last().second != i - 1) {
+                dataChangedRanges << qMakePair(i, i);
+            } else {
+                dataChangedRanges.last().second = i;
+            }
+        }
+    }
+
+    // Remaining new free timers
+    int inserted = 0;
+    for (auto it = changes.constBegin(), end = changes.constEnd(); it != end; ++it) {
+        // If a TimerId of type TimerId::QObjectType just changed then this free timer id is still valid,
+        // remove it from qtimerIds.
+        if (it.key().type() == TimerId::QObjectType && qtimerIds.contains(it.key().timerId())) {
+            qtimerIds.remove(it.key().timerId());
+        }
+
+        if (handledIds.contains(it.key()))
+            continue;
+
+        m_freeTimersInfo << it.value();
+        ++inserted;
+    }
+
+    // Inform model about new rows
+    if (inserted > 0) {
+        const int first = m_sourceModel->rowCount() + m_freeTimersInfo.count() - inserted;
+        const int last = m_sourceModel->rowCount() + m_freeTimersInfo.count() - 1;
+
+        beginInsertRows(QModelIndex(), first, last);
+        // We already inserted at end
```

Comment:
Does this still pass modeltest? This looks like it breaks some rowCount invariants in there.